akka-stream/src/main/scala/akka/stream/stage/Stage.scala

```diff
@@ -85,6 +87,22 @@ private[stream] abstract class AbstractStage[-In, Out, PushD <: Directive, PullD
    */
   def onUpstreamFailure(cause: Throwable, ctx: Ctx): TerminationDirective = ctx.fail(cause)
 
+  /**
+   * If an exception is thrown from [[#onPush]] this method is invoked to decide how
+   * to handle the exception. By default this method returns [[Supervision.Stop]].
+   *
+   * If an exception is thrown from [[#onPull]] the stream will always be completed with
+   * failure, because it is not always possible to recover from that state.
```

Comment:
maybe add some doc that the users should use their own try-catch-recover in onPull if they know how their logic is supposed to recover from that state.